[P159] Add LD2410 Radar presence detection

tonhuisman opened the pull request:
Resolves #4673 

P159 - Presence - LD2410 (24 GHz Radar human presence detector)

Using a _beta_ version of [nvmreynolds/ld2410](https://github.com/ncmreynolds/ld2410) library, with an open PR included to have Engineering mode, and other experimental patches like the Ambient Light sensor support.

Features:
- All presence related measurements available from the sensor
- All Engineering measurements available from the sensor
- Ambient Light value (undocumented feature in Engineering mode, only available on boards that have the AL sensor)
- Output pin state (undocumented feature in Engineering mode)
- Supported commands:
  - `ld2410,factoryreset` : Reset sensor to factory default settings (no questions asked!)
  - `ld2410,logall` : Write all available values at INFO log
- Available values: (usage: `[<taskName>#<valueName>]`
  - `Presence` : Presence (0/1)
  - `StatPres` : Stationary presence (0/1)
  - `MovPres` : Moving presence (0/1)
  - `Distance` : Distance in cm
  - `StatDist` : Stationary distance in cm
  - `MovDist` : Moving distance in cm
  - `StatEnergy` : Stationary object energy level
  - `MovEnergy` : Moving object energy level
  - `StatSensGate0`..`StatSensGate8` : Stationary gate sensitivity value for gate 0..8 (range 0..100)
  - `MovSensGate0`..`MovSensGate8` : Moving gate sensitivity value for gate 0..8 (range 0..100)
  - With 'Engineering mode' enabled:
  - `AmbLight` : Ambient light level (0..255) (unknown unit)
  - `OutputPin` : State of the output pin (0/1)
  - `StatEnergyGate0`..`StatEnergyGate8` : Stationary object energy level for gate 0..8, 75 cm/gate
  - `MovEnergyGate0`..`MovEnergyGate8` : Moving object energy for gate 0..8, 75 cm/gate
  - `StatSensGate0`..`StatSensGate8` : Stationary sensitivity setting gate 0..8
  - `MovSensGate0`..`MovSensGate8` : Moving sensitivity setting gate 0..8


TODO:
- [x] Rework fetching the data to make it respond faster to presence change, independent from Interval
- [x] Testing
  - [x] LD2410 ([confirmed](https://github.com/letscontrolit/ESPEasy/issues/4673#issuecomment-1784717824))
  - [x] LD2410B
  - [x] LD2410C ([confirmed](https://github.com/letscontrolit/ESPEasy/issues/4673#issuecomment-1784237246))
  - [x] Running from ESP8266 😊 ~(currently tested on ESP32 only)~
- [x] Documentation
- [x] Add configuration options for fine-tuning the detection settings
- [x] Enable current configuration settings to be presented in the UI
- [x] ... and used in rules
- [x] Add commands:
  - [x] FactoryReset
  - [x] LogAll
- [x] Update EasyCodeMirror

tonhuisman committed:
[P159] Add LD2410 Radar presence detection

TD-er commented on src/_P159_LD2410.ino:
```diff
@@ -0,0 +1,245 @@
+      P159_data_struct *P159_data = static_cast<P159_data_struct *>(getPluginTaskData(event->TaskIndex));
+
+      if (nullptr != P159_data) {
+        success = P159_data->processSensor(); // This has 50 msec delays included, so should NOT go in PLUGIN_FIFTY_PER_SECOND !
```
50 msec delay?
Sounds like we should split this into a set and read out call...
50 msec delay is not even working when you have 2 instances running.

tonhuisman replied:
I've been thinking about that, but didn't get to it yet, needs quite some work on the library. I did avoid the 2500 msec delay call during `begin()` though 👼

TD-er commented on src/src/PluginStructs/P159_data_struct.cpp:
```diff
@@ -0,0 +1,190 @@
+                                                bool    displayString) {
+  const __FlashStringHelper *strings[] { /*** ATTENTION: Don't change order of values as these are stored as user-selected!!! ***/
+    F("Presence"), F("Presence"),
+    F("Stationary Presence"), F("StationaryPresence"),
```
Quite long default taskvar names.
Can't these be shortened?

tonhuisman replied:
Yeah, if only to make it more readable.

TD-er commented on src/src/PluginStructs/P159_data_struct.h:
```diff
@@ -0,0 +1,98 @@
+// P159 "Dust - PM1006(K) (Vindriktning)"
+// Header file for sensor abstraction
+// 2022 By flashmark
+// ----------------------------------------------------------------------------
```
Ah busted :)
Copy/pasteing :)

tonhuisman committed:
[P159] Remove copy/paste evidence :-D

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P159-ld2410-presence-detection

tonhuisman committed:
[P159] Many improvements, add/update settings for sensitivity

TD-er commented on src/src/PluginStructs/P159_data_struct.cpp:
```diff
@@ -11,14 +6,14 @@ const __FlashStringHelper* Plugin_159_valuename(uint8_t value_nr,
-    F("Moving Object energy"), F("MovingEnergy"),
-    F("Ambient light sensor"), F("AmbientLight"),
+    F("Stationary Object distance"), F("StatDistance"),
+    F("Moving Object distance"), F("MovDistance"),
```
Why not `StatDist` and `MovDist` ?

TD-er commented on src/src/PluginStructs/P159_data_struct.cpp:
```diff
@@ -79,55 +74,73 @@ bool P159_data_struct::processSensor() {
     }
+
+    if (P159_state_e::Running != sState) { // FIXME Remove log
+      addLog(LOG_LEVEL_INFO, strformat(F("LD2410: Starting state: %d duration: %u msec."), static_cast<uint8_t>(sState), millis() - iStart));
```
`timePassedSince()` :)
And why limit this to an `uint8_t` ?

tonhuisman committed:
[P159] Fix compilation on LIMIT_BUILD_SIZE builds

tonhuisman committed:
[P159] Small improvements

tonhuisman committed:
[P159] Small improvements

tonhuisman committed:
[P159] Add command for sensor factoryreset

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P159-ld2410-presence-detection

tonhuisman committed:
[Lib] LD2410: Fix bug (-1 for uint), add undocumented OutputPinState value

tonhuisman committed:
[P159] Add OutputPinState option after AmbientLight option (breaking change!)

tonhuisman committed:
[P159] Minor change to determine how that may influence the build

tonhuisman committed:
[Build] Freeze Python to 3.11 to fix pygit2 build issue

tonhuisman committed:
[P159] Add Gate0 energy values that where missing
Add GetConfig support for fetching all available values
Add ld2410,logall command for dumping all values in the log (INFO level)
Code improvements

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P159-ld2410-presence-detection

tonhuisman committed:
[Lib] LD2410: Read all data that can be read in 5 msec, add errorcount and data_valid checks

tonhuisman committed:
[P159] Rework processing, now near-instant response
Add GetConfig support for sensitivity per gate values

TD-er commented on lib/ld2410/src/ld2410.cpp:
```diff
@@ -261,115 +261,134 @@ bool ld2410::read_frame_()
+		yield();
+		if(frame_started_ == false)
 		{
+			uint8_t byte_read_ = radar_uart_ -> read();
```
Why spaces here surrounding the `->` ?

tonhuisman replied:
That's all over that source, didn't do a source reformat (yet), as it will change _every_ line of these files...

TD-er commented on src/src/PluginStructs/P159_data_struct.cpp:
```diff
@@ -335,23 +374,39 @@ bool P159_data_struct::plugin_get_config_value(struct EventStruct *event,
-      if (logAll ||
-          cmd.equalsIgnoreCase(concat(Plugin_159_valuename(idx, false), (gate > -1) ? String(gate) : F("")))) {
+      if ((idx > -1) && (logAll ||
+                         cmd.equalsIgnoreCase(concat(Plugin_159_valuename(idx, false), (gate > -1) ? String(gate) : F(""))))) {
```
Only use the `?` ... `:` operator when both have the same return type.
`String` and `F()` are not the same type.

tonhuisman committed:
[Lib] LD2410: Reformat source (nearly all is changed!) Replace yield() by delay(0)

tonhuisman committed:
[P159] Fix wrong use of F(""), should be EMPTY_STRING

iz8mbw commented:
You can mark "Testing on LD2410C" as completed. Thanks Ton for you job!

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P159-ld2410-presence-detection

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P159-ld2410-presence-detection

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P159-ld2410-presence-detection

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P159-ld2410-presence-detection

tonhuisman committed:
[P159] Add documentation

iz8mbw commented:
@tonhuisman Ton please update also https://github.com/letscontrolit/ESPEasy/blob/mega/src/Custom-sample.h adding this new Plugin

tonhuisman commented:
Already added that: https://github.com/letscontrolit/ESPEasy/pull/4852/files#diff-7359af4349b63c67cab7f910f52d526808ab4e27aaa5901dc201e252f2196a61R524-R527 😄

iz8mbw commented:
Oh sorry, I miss it due to viewing with mobile phone

iz8mbw commented:
@tonhuisman as we know we well tested it on LD2410C and LD2410B models.
[Here](https://github.com/letscontrolit/ESPEasy/pull/4852/files#diff-7359af4349b63c67cab7f910f52d526808ab4e27aaa5901dc201e252f2196a61) you specify "LD24x0" so I suppose you know the Plugin is software compatible also with LD2420?
And also the newest LD2410S, LD2450, LD2411S and LD2411?
If yes then make sense to rename "Presence - LD2410" to "Presence - LD24x0x" if instead we don't know or we know it's not software compatible we should not specify it as "LD24x0".

iz8mbw commented:
On ESPHome it seems they have 2 different implementation between LD2410 and LD2420.

https://esphome.io/components/sensor/ld2410
https://esphome.io/components/sensor/ld2420

tonhuisman commented:
> [Here](https://github.com/letscontrolit/ESPEasy/pull/4852/files#diff-7359af4349b63c67cab7f910f52d526808ab4e27aaa5901dc201e252f2196a61) you specify "LD24x0" so I suppose you know the Plugin is software compatible also with LD2420?

Originally I intended to support both the LD2410 and LD2420, but AFAIR, they are not ~fully~ very software compatible. Especially the changed baud rate, depending on the firmware revision, sounds troublesome/confusing (for end-users) to me, but also the LD2420 has a lot less data values and configuration options available. Still not sure if I'll implement all these derived models, and I don't expect to do that in this PR, but in follow-up PR's. Similarly, I've already ordered a 60 GHz radar presence sensor, that has even different features, like fall-detection... 😉 

I'll have a look at what the differences between the ESPHome implementation implies, maybe it's needed there because of the change in available values, IDK.

iz8mbw commented:
OK, I totally agree with you, so if LD2410 and LD2420 are not very software compatible, it's better to leave this new Plugin as is it only for LD2410.
If you can also implement LD2420 it would be interesting.

About the other 60 GHz radar presence sensor, what is it? What chip?

Thanks!

iz8mbw commented:
and these new kind of Presence sensor using mmWave are much better than PIR.
mmWave technology is the future on presence detection.

iz8mbw commented:
[This](https://github.com/esphome/esphome/pull/4847) seems the LD2420 implementation in ESPHome.
See also: https://github.com/esphome/feature-requests/issues/2219

tonhuisman committed:
[P159] Update EasyColorMirror

tonhuisman committed:
[P159] Update comments to explicitly name LD2410

netson01 commented:
Where can I get build that supports LD2410?

[This](https://github.com/letscontrolit/ESPEasy/actions/runs/6882140118) test build has expired links.
Releases [Nov 30, 2023](https://github.com/letscontrolit/ESPEasy/releases/tag/mega-20231130) and [Dec 26, 2023](https://github.com/letscontrolit/ESPEasy/releases/tag/mega-20231225) have no "_Presence - LD2410 Radar detection_" in the device list.

Tested both releases on ESP8266 and ESP32 (normal_ESP8266_4M1M and normal_ESP32_4M316k).

tonhuisman commented:
@netson01 

> Tested both releases on ESP8266 and ESP32 (normal_ESP8266_4M1M and normal_ESP32_4M316k).

On the [Plugin overview page](https://espeasy.readthedocs.io/en/latest/Plugin/_Plugin.html) you can see that the plugin is included in the Collection G builds. If you install that type of build you will be able to select the LD2410.
It's been available since the November 2023 release.

starfish1107 commented:
I noticed considerable data flooding from this plugin to the mqtt broker - how can I reduce it?

TD-er commented:
You can act on the events sent by the task in the rules.
From there you can try to decimate the traffic, for example only to send if there was no presence for N amount of time.

Sending to MQTT can then be done via the publish command, or you copy the required values from the event to a dummy task and then call `taskRun` for that dummy task.
For this you should no longer send from the LD2410 task, but only from this dummy task to the MQTT controller.

TD-er commented:
By the way, it is better not to reply to closed issues or merged pull requests as not a lot of people will read them.
Better open a new issue.
However, this was a good example of some question to ask on the forum.
